llvm/tools/llvm-readobj/ELFDumper.cpp

```diff
@@ -6208,6 +6263,13 @@ void ELFDumper<ELFT>::forEachRelocationDo(
       Warn(RangeOrErr.takeError());
     }
     break;
+  case ELF::SHT_AARCH64_AUTH_RELR:
+    if (Obj.getHeader().e_machine != EM_AARCH64) {
+      this->reportUniqueWarning(
```

Comment:
we should drop this for non-AArch64. They may define the value for other purposes. Just use the default code path to handle an unknown section header type.

Reply:
I suppose a slightly different approach would be better if we have concerns regarding other targets defining the same section type for unrelated purposes. `ELFDumper<ELFT>::forEachRelocationDo` should only be called on sections which pass the check `isRelocationSec<LEFT>(Sec)`. So, we can change this check and make it take the target into account - as for now, it returns true for `SHT_AARCH64_AUTH_RESULT` unconditionally. After such change, we can just use an `assert(Obj.getHeader().e_machine == EM_AARCH64)` under `case ELF::SHT_AARCH64_AUTH_RESULT` in `ELFDumper<ELFT>::forEachRelocationDo` to ensure the contract is satisfied.

@MaskRay Please let me know your thoughts on this.

Reply:
I haven't checked, but if we can guarantee `assert(Obj.getHeader().e_machine == EM_AARCH64)`. It's better than `if (EM_AARCH64 && ELF::SHT_AARCH64_AUTH_RELR) do`

Reply:
It turns out that we can't assume that only relocation sections are passed to `forEachRelocationDo` - there are code paths when other section types also get there, for example, `printSectionHeaders -> printRelocationsHelper -> forEachRelocationDo`. So, we must break here on non-AArch64 emachine (the switch does not contain a default branch, so break looks OK for me).

I also added checks against emachine in other places where appropriate and moved code/tests to target-specific parts where needed. @MaskRay Please see f9103d548b61a60f41e7529ea357b6ee2b8a0f66 and let me know if it is OK.

Reply:
Testing `EHeader.e_machine == EM_AARCH64 && SType == ELF::SHT_AARCH64_AUTH_RELR` is reasonable.